Sources/MongoSwift/SDAM.swift

```diff
@@ -427,12 +432,25 @@ extension TopologyDescription {
     }
 
     internal func filterReplicaSetServers(
-        readPreference _: ReadPreference?,
+        readPreference: ReadPreference?,
         servers: [ServerDescription]
     ) -> [ServerDescription] {
+        let tagSets = readPreference?.tagSets ?? []
+        let matches = servers.filter { doTagsMatch(serverTags: $0.tags, tagSets: tagSets) }
         // TODO: Filter out servers staler than maxStalenessSeconds
-        // TODO: Select servers matching the tag_sets
-        // While waiting for the above to be implemented, this helper just returns the servers it was passed
-        servers
+        return matches
+    }
+
+    internal func doTagsMatch(serverTags: [String: String], tagSets: [BSONDocument]) -> Bool {
+        if tagSets.isEmpty {
+            return true
+        }
+        for tagSet in tagSets {
```

Comment:
I think the logic here needs to be a little different. the spec says:

> If the tag_sets list is not empty, then tag sets are tried in order until a tag set matches at least one candidate server. All candidate servers matching that tag set are eligible servers. Subsequent tag sets in the list are ignored.

currently, I think the list of matches above will contain any server that matches any tag set. if I am understanding the spec wording correctly, instead, we need to find the first tag set that matches any server, and then return all servers matching that tag set. 

here's the equivalent Rust logic for some inspiration: https://github.com/mongodb/mongo-rust-driver/blob/cc7a1c3046e3eac925dcaf85d1e4876bd2a897f8/src/sdam/description/topology/server_selection/mod.rs#L370-L386

it would be good to add a test around this behavior as well

Reply:
Yep, thanks!